README.md
## Jasypt

[![NPM version][npm-image]][npm-url]
[![build status][travis-image]][travis-url]
[![Test coverage][codecov-image]][codecov-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/jasypt.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/jasypt
[travis-image]: https://img.shields.io/travis/rickyes/jasypt.svg?style=flat-square
[travis-url]: https://travis-ci.org/rickyes/jasypt
[codecov-image]: https://codecov.io/gh/rickyes/jasypt/branch/master/graph/badge.svg
[codecov-url]: https://codecov.io/github/rickyes/jasypt?branch=master
[download-image]: https://img.shields.io/npm/dm/jasypt.svg?style=flat-square
[download-url]: https://npmjs.org/package/jasypt

org.jasypt.util.text.BasicTextEncryptor for Node.js

#### 背景
`Spring Boot` 集成 `jasypt` 对配置项进行加密，为了与 `Java` 体系保持一致，于是有了 `Jasypt.js`

#### 使用
**`SDK`**
``` js
'use strict';

const Jasypt = require('jasypt');
const jasypt = new Jasypt();
// 设置秘钥
jasypt.setPassword('G0CvDz7oJn60');
// 加密
const encryptMsg = jasypt.encrypt('admin');
// 解密
const decryptMsg = jasypt.decrypt(encryptMsg);
```

**`命令行`**
``` sh
$ jasypt -h

Usage: jasypt [options]

Options:
  -v, --version               output the version number
  -p, --password <pwd>        The secret key
  -e, --encrypt <msg>         Text to be encrypting
  -d, --decrypt <encryptMsg>  Text to be decrypting
  -h, --help                  output usage information

Examples:

  $ jasypt -p 0x1995 -e admin
  $ jasypt -p 0x1995 -d nsbC5r0ymz740/aURtuRWw==
```

#### 解密加密项 for Spring Boot
``` js
// 对配置项的ENC(xxx)进行解密
const data = {
  code: 42,
  test: {
    db: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)',
    pwd: {
      a: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)'
    },
    asad: {
      pwd: {
        str: 'str',
        host: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)',
        pwd: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)'
      }
    },
    items: [{
      user: 'user1',
      pwd: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)'
    }, {
      user: 'user2',
      pwd: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)'
    }, {
      user: 'user3',
      pwd: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)'
    }],
  }
};

const jasypt = new Jasypt();
jasypt.setPassword('P8dEw34TgvbY');
jasypt.decryptConfig(data);
```
``` js
// 解密出来的内容
const data = {
  code: 42,
  test: {
    db: 'admin',
    pwd: {
      a: 'admin'
    },
    asad: {
      pwd: {
        str: 'str',
        host: 'admin',
        pwd: 'admin'
      }
    },
    items: [{
      user: 'user1',
      pwd: 'admin'
    }, {
      user: 'user2',
      pwd: 'admin'
    }, {
      user: 'user3',
      pwd: 'admin'
    }],
  }
};
```


CHANGELOG.md
# Jasypt

### Version 1.0.3 / `2020-05-08`
- [[83d3f3b75d]](https://github.com/rickyes/jasypt/commit/83d3f3b75d) - Optimize command line printing details.

### Version 1.0.2 / `2020-04-22`
- [[3787b75c9c]](https://github.com/rickyes/jasypt/commit/3787b75c9c) - Use Buffer.from() and Buffer.alloc() to replace the new Buffer() that was deprecated.
- [[cd902dde01]](https://github.com/rickyes/jasypt/commit/cd902dde01) - add benchmark test

### Version 1.0.1 / `2019-01-04`
- [[e49b1f0ce5]](https://github.com/rickyes/jasypt/commit/e49b1f0ce5) - Repair default command output

test.js
'use strict';

const test = require('ava');

const Jasypt = require('.');
const password = 'G0CvDz7oJn60';
const message = 'admin';
const decryptMessage = 'c0KA89TBZ6TbLn7E6RIiFQ==';

const data = {
  code: 42,
  test: {
    db: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)',
    pwd: {
      a: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)'
    },
    asad: {
      pwd: {
        str: 'str',
        host: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)',
        pwd: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)'
      }
    },
    items: [{
      user: 'user1',
      pwd: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)'
    }, {
      user: 'user2',
      pwd: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)'
    }, {
      user: 'user3',
      pwd: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)'
    }],
  }
};

test('encrypt', t => {
  const jasypt = new Jasypt();
  jasypt.setPassword(password);
  const encryptMsg = jasypt.encrypt('');
  t.is(encryptMsg, null);
  jasypt.encrypt('a');
  t.pass();
});

test('decrypt', t => {
  const jasypt = new Jasypt();
  jasypt.setPassword(password);
  let decryptMsg = '';
  decryptMsg = jasypt.decrypt(null);
  t.is(decryptMsg, null);
  decryptMsg = jasypt.decrypt(decryptMessage);
  t.is(decryptMsg, message);
  t.pass();
});

test('encrypt & decrypt', t => {
  const jasypt = new Jasypt();
  jasypt.setPassword(password);
  const encryptMsg = jasypt.encrypt(message);
  const decryptMsg = jasypt.decrypt(encryptMsg);
  t.is(decryptMsg, message);
  t.pass();
});

test('decryptConfig', t => {
  const jasypt = new Jasypt();
  jasypt.setPassword(password);
  jasypt.decryptConfig('');
  jasypt.decryptConfig(data);
  t.is(data.test.db, message);
  t.is(data.test.pwd.a, message);
  t.is(data.test.asad.pwd.str, 'str');
  t.is(data.test.asad.pwd.pwd, message);
  t.is(data.test.asad.pwd.host, message);
  for (const item of data.test.items) {
    t.is(item.pwd, message);
  }
  t.pass();
});




test/jasypt.test.js
'use strict';

const {test} = require('tap');

const Jasypt = require('..');
const password = 'G0CvDz7oJn60';
const message = 'admin';
const decryptMessage = 'c0KA89TBZ6TbLn7E6RIiFQ==';

const data = {
  code: 42,
  test: {
    db: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)',
    pwd: {
      a: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)'
    },
    asad: {
      pwd: {
        str: 'str',
        host: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)',
        pwd: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)'
      }
    },
    items: [{
      user: 'user1',
      pwd: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)'
    }, {
      user: 'user2',
      pwd: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)'
    }, {
      user: 'user3',
      pwd: 'ENC(c0KA89TBZ6TbLn7E6RIiFQ==)'
    }],
  }
};

test('encrypt', t => {
  const jasypt = new Jasypt();
  jasypt.setPassword(password);
  const encryptMsg = jasypt.encrypt('');
  t.equal(encryptMsg, null);
  jasypt.encrypt('a');
  t.end();
});

test('decrypt', t => {
  const jasypt = new Jasypt();
  jasypt.setPassword(password);
  let decryptMsg = '';
  decryptMsg = jasypt.decrypt(null);
  t.equal(decryptMsg, null);
  decryptMsg = jasypt.decrypt(decryptMessage);
  t.equal(decryptMsg, message);
  t.end();
});

test('encrypt & decrypt', t => {
  const jasypt = new Jasypt();
  jasypt.setPassword(password);
  const encryptMsg = jasypt.encrypt(message);
  const decryptMsg = jasypt.decrypt(encryptMsg);
  t.equal(decryptMsg, message);
  t.end();
});

test('decryptConfig', t => {
  const jasypt = new Jasypt();
  jasypt.setPassword(password);
  jasypt.decryptConfig('');
  jasypt.decryptConfig(data);
  t.equal(data.test.db, message);
  t.equal(data.test.pwd.a, message);
  t.equal(data.test.asad.pwd.str, 'str');
  t.equal(data.test.asad.pwd.pwd, message);
  t.equal(data.test.asad.pwd.host, message);
  for (const item of data.test.items) {
    t.equal(item.pwd, message);
  }
  t.end();
});




encryptor.js
'use strict';

const assert = require('assert');
const crypto = require('crypto');
const util = require('./util');

class Encryptor {
  constructor() {
    this.algorithm = '';
  }

  /**
   * 设置加密算法
   * @param {String} algorithm 算法
   */
  setAlgorithm(algorithm) {
    assert(!util.isEmpty(algorithm), `Algorithm cannot be set empty: ${algorithm}`);
    this.algorithm = algorithm;
  }

  /**
   * 设置偏移量
   * @param {String} password 秘钥
   * @param {String} salt 盐
   * @param {String} iterations 偏移量
   */
  KDF (password, salt, iterations) {
    const pwd = Buffer.from(password, 'utf-8');
    let key = Buffer.concat([pwd, salt]);
    for (let i = 0; i < iterations; i++) {
      key = crypto.createHash('md5').update(key).digest();
    }
    return key;
  }

  /**
   * 获取秘钥和盐
   * @param {String} password 秘钥
   * @param {String} salt 盐
   * @param {String} iterations 偏移量
   */
  getKeyIV(password, salt, iterations) {
    const key = this.KDF(password, salt, iterations);
    const keybuf = Buffer.from(key, 'binary').slice(0, 8);
    const ivbuf = Buffer.from(key, 'binary').slice(8, 16);
    return [ keybuf, ivbuf ];
  }

  /**
   * 加密
   * @param {String} payload 需要加密的文本
   * @param {String} password 秘钥
   * @param {String} salt 盐
   * @param {String} iterations 偏移量
   */
  encrypt(payload, password, salt, iterations) {
    const kiv = this.getKeyIV(password, salt, iterations);
    const cipher = crypto.createCipheriv('des', kiv[0], kiv[1]);

    const encrypted = [];
    encrypted.push(cipher.update(payload, 'utf-8', 'hex'));
    encrypted.push(cipher.final('hex'));

    const out = Buffer.from(encrypted.join(''), 'hex');
    const result =  Buffer.alloc(out.length + salt.length);

    salt.copy(result, 0, 0, salt.length);
    out.copy(result, salt.length, 0, out.length);

    return result.toString('base64');
  }

  /**
   * 解密
   * @param {String} payload 需要解密的文本
   * @param {String} password 秘钥
   * @param {String} iterations 偏移量
   */
  decrypt(payload, password, iterations) {
    const encryptedMessage = Buffer.from(payload, 'base64');
    const decrypted = [];
    const saltStart = 0;
    const saltSizeBytes = 8;

    const saltSize = saltSizeBytes < encryptedMessage.length ? saltSizeBytes : encryptedMessage.length;
    const encMesKernelStart = saltSizeBytes < encryptedMessage.length ? saltSizeBytes : encryptedMessage.length;
    const encMesKernelSize = saltSizeBytes < encryptedMessage.length ? (encryptedMessage.length - saltSizeBytes) : 0;

    const salt = Buffer.alloc(saltSize);
    const encryptedMessageKernel = Buffer.alloc(encMesKernelSize);

    encryptedMessage.copy(salt, 0, saltStart, saltSize);
    encryptedMessage.copy(encryptedMessageKernel, 0, encMesKernelStart, encryptedMessage.length);

    const kiv = this.getKeyIV(password, salt, iterations);
    const decipher = crypto.createDecipheriv('des', kiv[0], kiv[1]);

    decrypted.push(decipher.update(encryptedMessageKernel));
    decrypted.push(decipher.final());

    return decrypted.join('');
  }
}

module.exports = Encryptor;

benchmark/jasypt.js
'use strict';

const Benchmark = require('benchmark');
const benchmarks = require('beautify-benchmark');
const suite = new Benchmark.Suite();
const Jasypt = require('..');

const jasypt = new Jasypt();
jasypt.setPassword('0x1995');

const data = {
  code: 42,
  test: {
    db: 'ENC(ok0kgkHP0Zj3WlYX4MA12w==)',
    pwd: {
      a: 'ENC(ok0kgkHP0Zj3WlYX4MA12w==)'
    },
  },
};

suite
.add('encrypt', function () {
  jasypt.encrypt('admin');
})
.add('decrypt', function () {
  jasypt.decrypt('ok0kgkHP0Zj3WlYX4MA12w==');
})
.add('decryptConfig', function () {
  jasypt.decryptConfig(data);
  data.test.db = 'ENC(ok0kgkHP0Zj3WlYX4MA12w==)';
  data.test.pwd.a = 'ENC(ok0kgkHP0Zj3WlYX4MA12w==)';
})
.on('cycle', function (event) {
  benchmarks.add(event.target);
})
.on('start', function (event) {
  console.log('\n  map Benchmark\n  node version: %s, date: %s\n  Starting...',
    process.version, Date());
})
.on('complete', function done() {
  benchmarks.log();
})
.run({ 'async': false });